driver/utils/type_info.h

```diff
@@ -0,0 +1,2112 @@
+    using SimpleTypeWrapper<std::uint8_t>::SimpleTypeWrapper;
+};
+
+template <>
+struct DataSourceType<DataSourceTypeId::UInt16>
+    : public SimpleTypeWrapper<std::uint16_t>
+{
+    using SimpleTypeWrapper<std::uint16_t>::SimpleTypeWrapper;
+};
+
+template <>
+struct DataSourceType<DataSourceTypeId::UInt32>
+    : public SimpleTypeWrapper<std::uint32_t>
+{
+    using SimpleTypeWrapper<std::uint32_t>::SimpleTypeWrapper;
+};
+
+template <>
+struct DataSourceType<DataSourceTypeId::UInt64>
+    : public SimpleTypeWrapper<std::uint64_t>
+{
+    using SimpleTypeWrapper<std::uint64_t>::SimpleTypeWrapper;
+};
+
+template <>
+struct DataSourceType<DataSourceTypeId::UUID>
+    : public SimpleTypeWrapper<SQLGUID>
+{
+    using SimpleTypeWrapper<SQLGUID>::SimpleTypeWrapper;
+};
+
+template <class T> struct is_string_data_source_type
+    : public std::false_type
+{
+};
+
+template <> struct is_string_data_source_type<DataSourceType<DataSourceTypeId::String>>
+    : public std::true_type
+{
+};
+
+template <> struct is_string_data_source_type<DataSourceType<DataSourceTypeId::FixedString>>
+    : public std::true_type
+{
+};
+
+template <class T> inline constexpr bool is_string_data_source_type_v = is_string_data_source_type<T>::value;
+
+// Used to avoid duplicate specializations in platforms where 'std::int32_t' or 'std::int64_t' are typedef'd as 'long'.
+struct long_if_not_typedefed {
+    struct dummy {};
+    using type = std::conditional_t<
+        (std::is_same_v<long, std::int32_t> || std::is_same_v<long, std::int64_t>),
+        dummy,
+        long
+    >;
+};
+
+// Used to avoid duplicate specializations in platforms where 'std::uint32_t' or 'std::uint64_t' are typedef'd as 'unsigned long'.
+struct unsigned_long_if_not_typedefed {
+    struct dummy {};
+    using type = std::conditional_t<
+        (std::is_same_v<unsigned long, std::uint32_t> || std::is_same_v<unsigned long, std::uint64_t>),
+        dummy,
+        unsigned long
+    >;
+};
+
+namespace value_manip {
+
+    template <typename T>
+    inline std::int16_t getColumnSize(const T & obj, const TypeInfo & type_info) {
+        return type_info.column_size;
+    }
+
+    inline std::int16_t getColumnSize(const SQL_NUMERIC_STRUCT & numeric, const TypeInfo & type_info) {
+        return (numeric.precision == 0 ? type_info.column_size : numeric.precision);
+    }
+
+    // TODO: implement getColumnSize() for other types.
+
+    template <typename T>
+    inline std::int16_t getDecimalDigits(const T & obj, const TypeInfo & type_info) {
+        return 0;
+    }
+
+    inline std::int16_t getDecimalDigits(const SQL_NUMERIC_STRUCT & numeric, const TypeInfo & type_info) {
+        return numeric.scale;
+    }
+
+    // TODO: implement getDecimalDigits() for other types.
+
+    template <typename ProxyType, typename SourceType, typename DestinationType>
+    void convert_via_proxy(const SourceType & src, DestinationType & dest);
+
+    template <typename SourceType>
+    struct from_value {
```

Comment:
IMHO, `from_value<T>::to_value<Y>::convert()` is a bit of over complication (both in calling code and implementation), could that be set of `convertFromValue(const X & x, Y & y)` function overloads ?
Also, that would allow moving all that code into `.cpp`

Reply:
During the evolution of this code I went through different configurations, and I actually saw what you mentioned too. It was just too non-uniform and hard to maintain, to have such different ways of calling the conversion code at that time. Things become more complicated when you try to do SFINAE or partial specializations. So I picked this representation, because it is actually the only maintainable one.

Reply:
This can be done in a more simple manner: https://godbolt.org/z/Vxke_T

Reply:
Maybe... if it could be applied to the entire code, including partial specialization cases.